### PR TITLE
Mathlib Vectors

### PR DESCRIPTION
**Pull request by pampersrocker**
Adds a default Vector3 implementation ( `krepel.math.vector3.Vector3` ) to the `krepel.math` module.
A simd version of this code may be created later in a module e.g. `krepel.math.simd`

Progress:
- [x] struct specification
- [x] Initialization
- [x] Addition
- [x] Subtraction
- [x] Dot Product
- [x] Cross Product
- [x] Float Scaling
- [x] Operators
- [x] Normalization
- [x] NaN checks
- [x] Default Values (Forward, Right, Up , UnitScale, Zero)
- [x] Normal Projection
- [x] Plane Projection
- [x] Normal Mirroring
- [x] Unit tests for everything
- [x] Documentation
- [x] Clamping
- [x] Swizzle Operator (dispatch operator?)
- [x] Nearly Equals

### CODE REVIEW

File: code/krepel/math/vector3.d [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+        "Y" ~ op ~ "rhs,"
+        "Z" ~ op ~ "rhs)");
+    }
+    else
+    {
+      assert(false, "Operator " ~ op ~ " not implemented.");
+    }
+  }
+
+  static Vector3 Forward = Vector3(1,0,0);
```
Comment (Manuzor) [inline code/krepel/math/vector3.d]: These should be `__gshared immutable Foo = Vector3(...);`.

(`immutable` and `const` act as if you stated `auto`, so there's no need to state the type twice.)

Comment (pampersrocker) [inline code/krepel/math/vector3.d]: Making them `immutable` also revealed that the operators had no `const` overloading ;)

**Comment by pampersrocker** [comment]
Everything should be fine and dandy now.

File: code/krepel/math/math.d [hunk 2]
```diff
@@ -0,0 +1,82 @@
+module krepel.math.math;
+
+import std.math;
+import krepel.algorithm.comparison;
+
+@safe:
+@nogc:
```
Comment (Manuzor) [inline code/krepel/math/math.d:7]: You could also add `nothrow`.

File: code/krepel/math/math.d [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    xor AH, 0x40; // Check if ZF is set and CF is zero.
+    mov result, AH; // Write result into bool variable
+  }
+  return !result;
+}
+
+/// Checks if a and b are equals with respect to some epsilon
+/// Epsilon gets scaled by the bigger of the two values to compare
+/// to compensate the reduced precision when using higher magnitudes
+bool NearlyEquals(float a, float b, float epsilon = 1.0e-4f)
```
Comment (Manuzor) [inline code/krepel/math/math.d]: Do we want a global constant that defines the default epsilon? Such as `enum KindaSmallNumber = 1.0e-4f;`? Or like in ezEngine with something like this:

```
enum : float
{
  HugeEpsilon    = 1.0e-3f,
  LargeEpsilon   = 1.0e-4f,
  DefaultEpsilon = 1.0e-5f,
  SmallEpsilon   = 1.0e-6f,
}
```

I'm fine with hard-coding the epsilon though.

File: code/krepel/math/math.d [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+  assert(NearlyEquals(d,e));
+  assert(!NearlyEquals(d,f));
+  assert(!NearlyEquals(d,g));
+}
+
+/// NaN Test
+unittest
+{
+  float f;
+  float correct =1.0f;
```
Comment (Manuzor) [inline code/krepel/math/math.d]: Ehrmagherd, no space after the `=` :scream:

File: code/krepel/math/vector3.d [hunk 5]
```diff
@@ -0,0 +1,479 @@
+module krepel.math.vector3;
+
+import krepel.math.math;
+import krepel.algorithm.comparison;
+import std.conv;
+
+@nogc:
+@safe:
```
Comment (Manuzor) [inline code/krepel/math/vector3.d:10]: Could add `nothrow` (again).

File: code/krepel/math/vector3.d [hunk 6]
```diff
@@ -0,0 +1,479 @@
...
+/// Calculates the squared length  of the given vector, which is the same as the dot product with the same vector
+/// Input vector will not be modified
+float LengthSquared(Vector3 vec)
+{
+  return vec | vec;
+}
+
+/// Calculates the squared length of the X and Y components, ignoring the Z component
+/// Input vector will not be modified
+float LengthSquared2D(Vector3 vec)
```
Comment (Manuzor) [inline code/krepel/math/vector3.d]: Maybe the `-2D` version become obsolete with the swizzling operators? `v3.LengthSquared2D()` vs. `v3.XY.LengthSquared()`.

File: code/krepel/math/vector3.d [hunk 7]
```diff
@@ -0,0 +1,479 @@
...
+
+  this(float X, float Y, float Z)
+  {
+    this.X = X;
+    this.Y = Y;
+    this.Z = Z;
+  }
+
+  /// Normalizes the vector (vector will have a length of 1.0)
+  void Normalize()
```
Comment (Manuzor) [inline code/krepel/math/vector3.d:174]: :+1: for not returning the result!

File: code/krepel/math/vector3.d [hunk 8]
```diff
@@ -0,0 +1,479 @@
...
+      auto result = mixin("Vector3("~
+        "X" ~ op ~ "rhs,"
+        "Y" ~ op ~ "rhs,"
+        "Z" ~ op ~ "rhs)");
+      Data = result.Data;
+      return this;
+    }
+    else
+    {
+      assert(false, "Operator " ~ op ~ " not implemented.");
```
Comment (Manuzor) [inline code/krepel/math/vector3.d]: Should be `static assert(...)`

File: code/krepel/math/vector3.d [hunk 9]
```diff
@@ -0,0 +1,479 @@
...
+    {
+      return Vector3(
+        (Y * rhs.Z) - (Z * rhs.Y),
+        (Z * rhs.X) - (X * rhs.Z),
+        (X * rhs.Y) - (Y * rhs.X)
+      );
+    }
+    else
+    {
+      assert(false, "Operator " ~ op ~ " not implemented.");
```
Comment (Manuzor) [inline code/krepel/math/vector3.d]: Should be `static assert(...)`

File: code/krepel/math/vector3.d [hunk 10]
```diff
@@ -0,0 +1,479 @@
...
+{
+  return krepel.math.NearlyEquals(a.X, b.X, epsilon) &&
+         krepel.math.NearlyEquals(a.Y, b.Y, epsilon) &&
+         krepel.math.NearlyEquals(a.Z, b.Z, epsilon);
+}
+
+/// Returns a clamped copy of the given vector
+/// The retuned vector will be of size <= MaxSize
+/// Input vector will not be modified
+Vector3 ClampSize(Vector3 vec, float MaxSize)
```
Comment (Manuzor) [inline code/krepel/math/vector3.d]: For the same reason `Normalize` and `NormalizedCopy` exist, shouldn't this be named similarly instead?

`SizeClampedCopy`, `ClampSizeCopy`, `ClampedToSizeCopy`, ... they all sound weird though :stuck_out_tongue:

Maybe just `ClampedSize` for those that copy and `ClampSize` for those that do it in-place?

**Comment by Manuzor** [comment]
You often write your local variables / parameters in camelCaseWords, and some in actual PascalCase. Shouldn't all be written in PascalCase? I don't understand the pattern.

**Comment by pampersrocker** [comment]
This merge request is now here for all Vector implementations as they are now too intertwined to handle them separately.

File: code/krepel/math/vector3.d (hunk 10 above)
Comment (pampersrocker) [inline code/krepel/math/vector3.d]: Well there are not such things other than Normalize, all other operations create a copy.

File: code/krepel/math/math.d (hunk 4 above)
Comment (pampersrocker) [inline code/krepel/math/math.d]: :scream:

File: code/krepel/math/math.d (hunk 2 above)
Comment (pampersrocker) [inline code/krepel/math/math.d:7]: Done

**Comment by pampersrocker** [comment]
Swizzle operator is now fully functional, every vector can now be swizzled regarding the following regex: 

``` ruby
vec2 = /(_[0XY]{2,4}|[XY]{1}[0XY]{1,3})/
vec3 = /(_[0XYZ]{2,4}|[XYZ]{1}[0XYZ]{1,3})/
vec4 = /(_[0XYZW]{2,4}|[XYZW]{1}[0XYZW]{1,3})/
```

File: code/krepel/math/vector3.d [hunk 11]
```diff
@@ -0,0 +1,569 @@
...
+float LengthSquared(Vector3 Vec)
+{
+  return Vec | Vec;
+}
+
+/// Calculates the squared length of the X and Y components, ignoring the Z component
+/// Input Vector will not be modified
+float LengthSquared2D(Vector3 Vec)
+{
+  return Vec.X * Vec.X + Vec.Y * Vec.Y;
```
Comment (Manuzor) [inline code/krepel/math/vector3.d:47]: This is equivalent to `Vec.XY | Vec.XY` now, right? :smile:

File: code/krepel/math/vector3.d [hunk 12]
```diff
@@ -0,0 +1,569 @@
...
+      Data = Result.Data;
+      return this;
+    }
+    else
+    {
+      static assert(false, "Operator " ~ Operator ~ " not implemented.");
+    }
+  }
+
+  //const (char)[] ToString() const
```
Comment (Manuzor) [inline code/krepel/math/vector3.d:208]: I think we will do these kinds of things in a much more general way using compile-time reflection at some other point.

File: code/krepel/math/vector3.d [hunk 13]
```diff
@@ -0,0 +1,569 @@
...
+  }
+
+  // Vector3 Swizzle
+  unittest
+  {
+    assert(Vector3(1, 2, 3).ZX   == Vector2(3, 1));
+    assert(Vector3(1, 2, 3).XZX  == Vector3(1, 3, 1));
+    assert(Vector3(1, 2, 3).XYXY == Vector4(1, 2, 1, 2));
+    assert(Vector3(1, 2, 3)._ZYX == Vector3(3, 2, 1));
+    assert(Vector3(1, 2, 3)._0YX == Vector3(0, 2, 1));
```
Comment (Manuzor) [inline code/krepel/math/vector3.d:564]: Add a test for `0` inbetween without `_`? Like `Vector3(1, 2, 3).X0Z == Vector3(1, 0, 3)`.

File: code/krepel/math/vector3.d (hunk 11 above)
Comment (pampersrocker) [inline code/krepel/math/vector3.d:47]: Theoretically yes, but then the compiler cannot optimize this:

Dissasembly not using swizzle (note the SIMD instructions):

```
    42: 
    43: /// Calculates the squared length of the X and Y components, ignoring the Z component
    44: /// Input Vector will not be modified
    45: float LengthSquared2D(Vector3 Vec)
00007FF733460E70  push        rbp  
00007FF733460E71  mov         rbp,rsp  
00007FF733460E74  mov         qword ptr [Vec],rcx  
    46: {
    47:   return Vec.X * Vec.X + Vec.Y * Vec.Y;
00007FF733460E78  mov         rax,qword ptr [Vec]  
00007FF733460E7C  movss       xmm0,dword ptr [rax]  
00007FF733460E80  movss       xmm1,dword ptr [rax]  
00007FF733460E84  mulss       xmm0,xmm1  
00007FF733460E88  lea         rcx,[rax+4]  
00007FF733460E8C  movss       xmm2,dword ptr [rcx]  
00007FF733460E90  movss       xmm3,dword ptr [rcx]  
00007FF733460E94  mulss       xmm2,xmm3  
00007FF733460E98  addss       xmm0,xmm2  
    48: }
00007FF733460E9C  pop         rbp  
00007FF733460E9D  ret  
```

And using swizzle:

```
    42: 
    43: /// Calculates the squared length of the X and Y components, ignoring the Z component
    44: /// Input Vector will not be modified
    45: float LengthSquared2D(Vector3 Vec)
00007FF6FC66C140  push        rbp  
00007FF6FC66C141  mov         rbp,rsp  
00007FF6FC66C144  sub         rsp,10h  
00007FF6FC66C148  mov         qword ptr [Vec],rcx  
    46: {
    47:   return Vec.XY | Vec.XY;
00007FF6FC66C14C  mov         rcx,qword ptr [Vec]  
00007FF6FC66C150  sub         rsp,20h  
00007FF6FC66C154  call        _D6krepel4math7vector37Vector327__T10opDispatchVAyaa2_5859Z10opDispatchMNgFNbNiNfZS6krepel4math7vector27Vector2 (07FF6FC666311h)  
00007FF6FC66C159  add         rsp,20h  
00007FF6FC66C15D  mov         qword ptr [rbp-10h],rax  
00007FF6FC66C161  lea         rax,[rbp-10h]  
00007FF6FC66C165  mov         qword ptr [rbp-8],rax  
00007FF6FC66C169  mov         rcx,qword ptr [Vec]  
00007FF6FC66C16D  sub         rsp,20h  
00007FF6FC66C171  call        _D6krepel4math7vector37Vector327__T10opDispatchVAyaa2_5859Z10opDispatchMNgFNbNiNfZS6krepel4math7vector27Vector2 (07FF6FC666311h)  
00007FF6FC66C176  add         rsp,20h  
00007FF6FC66C17A  mov         rdx,rax  
00007FF6FC66C17D  mov         rcx,qword ptr [rbp-8]  
00007FF6FC66C181  sub         rsp,20h  
00007FF6FC66C185  call        _D6krepel4math7vector27Vector223__T8opBinaryHVAyaa1_7cZ8opBinaryMFNaNbNiNfS6krepel4math7vector27Vector2Zf (07FF6FC6650D3h)  
00007FF6FC66C18A  add         rsp,20h  
    48: }
00007FF6FC66C18E  mov         rsp,rbp  
00007FF6FC66C191  pop         rbp  
00007FF6FC66C192  ret  
```

Comment (pampersrocker) [inline code/krepel/math/vector3.d:47]: Also note that the here used simd instructions are far from optimal, but this is a task for the extra simd math.

**Comment by Manuzor** [comment]
Is there a way to do a safe normalization of a vector, i.e. still succeed when it's a zero-vector or the length is very close to zero?

**Pull request merged**